pkg/adapter/nimbus-kyverno/watcher/kpwatcher.go

```diff
@@ -39,42 +41,67 @@ func kpInformer() cache.SharedIndexInformer {
 
 // WatchKps watches update and delete event for KyvernoPolicies owned by
 // NimbusPolicy or ClusterNimbusPolicy and put their info on respective channels.
-func WatchKps(ctx context.Context, addKpch, updatedKpCh, deletedKpCh chan common.Request) {
+func WatchKps(ctx context.Context, updatedKpCh, deletedKpCh chan common.Request) {
 	logger := log.FromContext(ctx)
 	informer := kpInformer()
 	handlers := cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			u := obj.(*unstructured.Unstructured)
-			if adapterutil.IsOrphan(u.GetOwnerReferences(), "NimbusPolicy") {
-				logger.V(4).Info("Ignoring orphan KyvernoPolicy", "KyvernoPolicy.Name", u.GetName(), "KyvernoPolicy.Namespace", u.GetNamespace(), "Operation", "Delete")
-				return
-			}
-			kpNamespacedName := common.Request{
-				Name:      u.GetName(),
-				Namespace: u.GetNamespace(),
-			}
-			if strings.Contains(kpNamespacedName.Name, "mutateexisting") {
-				addKpch <- kpNamespacedName
-			}
-		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
 			oldU := oldObj.(*unstructured.Unstructured)
 			newU := newObj.(*unstructured.Unstructured)
-
 			if adapterutil.IsOrphan(newU.GetOwnerReferences(), "NimbusPolicy") {
 				logger.V(4).Info("Ignoring orphan KyvernoPolicy", "KyvernoPolicy.Name", oldU.GetName(), "KyvernoPolicy.Namespace", oldU.GetNamespace(), "Operation", "Update")
 				return
 			}
 
-			if oldU.GetGeneration() == newU.GetGeneration() {
+			var oldKp kyvernov1.Policy
+			var newKp kyvernov1.Policy
+			err := runtime.DefaultUnstructuredConverter.FromUnstructured(oldU.Object, &oldKp)
+			if err != nil {
+				logger.Error(err, "failed to convert to kyverno policy")
 				return
 			}
 
-			kpNamespacedName := common.Request{
-				Name:      newU.GetName(),
-				Namespace: newU.GetNamespace(),
+			err = runtime.DefaultUnstructuredConverter.FromUnstructured(newU.Object, &newKp)
+			if err != nil {
+				logger.Error(err, "failed to convert to kyverno policy")
+				return
+			}
+
+			// oldConditions := oldKp.Status.Conditions
+			newConditions := newKp.Status.Conditions
+			// oldConditionsSize := len(oldConditions)
+			// newConditionsSize := len(newConditions)
+
+			if !strings.Contains(newKp.GetName(), "mutateexisting") {
+				if oldU.GetGeneration() == newU.GetGeneration() {
+					return
+				}
+				kpNamespacedName := common.Request{
+					Name:      newU.GetName(),
+					Namespace: newU.GetNamespace(),
+				}
+				updatedKpCh <- kpNamespacedName
+				return
+			}
+
+			// for mutate existing policy
+			if oldU.GetGeneration() == newU.GetGeneration() {
+				if checkIfReady(newConditions) {
```

Comment:
The if should be: If checkIfReady(newConditions) && !checkIfReady(oldConditions) {